_posts/2024-01-29-the-things-i-like-about-cljs-and-re-frame.md

```diff
@@ -0,0 +1,354 @@
+![History](/img/the-things-i-like-about-cljs-and-re-frame/tuhoaja666-2.png)
+
+These are main tools that I've used and their purpose in a nutshell.
+
+### [ClojureScript](https://clojurescript.org/)
+
+ClojureScript is basically a compiler for compiling Clojure into JavaScript. Uses Google Closure for optimization.
+
+### [Leiningen](https://leiningen.org/)
+
+Tool for Clojure project management. In this case used only for creating the project using a template [Re-Frame -template project](https://github.com/day8/re-frame-template).
+A rather nifty way to get a working project template straight out of the box.
+
+### [Clojure-CLI (deps.edn)](https://clojure.org/guides/deps_and_cli)
+
+ClojureScript libraries and dependencies are handled using deps.edn in this project.
+
+### [Shadow-CLJS](https://github.com/thheller/shadow-cljs)
+
+ClojureScript build tool used in this project. Handles integration with NPM, live reload, REPL etc.
+
+### [Reagent](https://github.com/reagent-project/reagent)
+
+Simple interface for ClojureScript to React. Enables writing React-components using ClojureScript functions. Instead of the Reacts "not quite HTML" uses Hiccup, which is representated via Clojure basic datastructures (maps and vectors).
+
+### [Re-Frame](https://github.com/day8/re-frame)
+
+ClojureScript frontend framework. "Last in the chain", meaning it builds on Reagent and enables building React components. Alters the React basic paradigm a bit. Basically Re-Frame events are the only means for mutating the app state and views react to them via subscriptions.
+
+### [Stylefy](https://github.com/Jarzka/stylefy)
+CLJS-library made by our very own Jari Hanhela, which enables writing CSS styles as Clojure data and attaching them to Reagent components.
+
+### [Velho DS](https://github.com/velho-allianssi/velho-ds)
+Reagent component library developed by the Velho alliance.
+
+
+## Basic concepts
+
+Or maybe I should refer this chapters as scarcity of them. This is one of the main selling points of this setup for me. Basically the only things you need to know are the following:
+
+#### 1. App state 
+Also known as "db" with Re-Frame. Re-Frame takes all the app state and stores it in a so-called big atom. The main issue I had working with just Reagent was that it doesn't take long for the application state to blow up all over the place, making it development and testing a real pain. 
+With re-frame, you have it neatly stored in a single place.
+
+#### 2. Subscriptions
+To put it short, subscriptions simply are a way to get the state out to your views. They react to changes in the appstate. They can be composed or calculated on the fly.
+
+#### 3. Events
+Whereas subscriptions are away of getting the state out, events are away of mutating the state.
+
+#### 4. Views
+Basically React-components written using [Hiccup](https://github.com/weavejester/hiccup). These components get react to subscriptions and mutate the state via events.
```

Comment:
En ihan hoksaa tämän lauseen pointtia "These components get react to subscriptions and mutate the state via events.". Voisi koittaa jotenkin selventää.

Reply:
Ah helkkarin lukihäiriöpää. 👍